### ISSUE TITLE
"file ... does not exist in binary cache" seems to go on forever

### ISSUE THREAD

**Issue by ivan** [NixOS/nix#3002]
This happens with both nix 2.2.2 and 2.3pre6779_324a5dc9 on NixOS master, and even after doing `rm -rf /root/.cache/nix`. The machine has a binary cache pointed at itself (http nix-serve), which may be odd, but is a result of shared server configuration.

```
root@xssd:/root# nix --version
nix (Nix) 2.2.2

root@xssd:/root# rm -rf /root/.cache/nix

root@xssd:/root# nix-env -f /var/nixpkgs -iA chromium konsole
installing 'chromium-75.0.3770.142'
installing 'konsole-19.04.1'
these derivations will be built:
  /nix/store/g22kvv5csgjha330r55dhxqyabfifd1s-qtx11extras-5.12.0.drv
  /nix/store/4n6d5l7d4wydmmfl63kyvyycyqbncx69-qtsvg-5.12.0.drv
  /nix/store/lqrh0vlwhlryklj69f67iy67sdi9avis-qtdeclarative-5.12.0.drv
  /nix/store/l4nsjzxfpb3akzdk08560dx2qazdy8vd-qtquickcontrols-5.12.0.drv
  /nix/store/9b510cpp2k8rj4xjlzy3r13i2n1b3h8h-qtwayland-5.12.0.drv
  /nix/store/x2q7rpfr0i1iww64bp3r3mlds8mnnprx-hook.drv
  /nix/store/0yl1531z7bzsga74hzimmyvbj7srzf97-kwindowsystem-5.58.0.drv
  /nix/store/kcl9n7hbs0xsf2s554ipwphksad3bi2f-kcoreaddons-5.58.0.drv
  /nix/store/1w4706c2w6mdmz8wi0s8kg341pn948gc-kcrash-5.58.0.drv
  /nix/store/g8sg5ynzaqpkqhwvsvzsdqbk44hv4a6p-karchive-5.58.0.drv
  /nix/store/p0wd2451ywsj496xlhfgz2qjza5lxnq7-ki18n-5.58.0.drv
  /nix/store/854dn5k74zl27hix44mmfchi4713sgz2-kdoctools-5.58.0.drv
  /nix/store/f078jw6vdxyvwvyk5g9ns7pncp8rjlv2-kconfig-5.58.0.drv
  /nix/store/k1rm0ssnybil66qrnfb2jssm5fm74jjw-kdbusaddons-5.58.0.drv
  /nix/store/694gh69q3n45n5wiaqry9dcbs1ckv94y-kservice-5.58.0.drv
  /nix/store/2xjraz3xqi2axhjc067j80jv28hxw98s-kemoticons-5.58.0.drv
  /nix/store/4p7qninjj2w5lbgiisg5wz3n6frpqhxy-kauth-5.58.0.drv
  /nix/store/60w8qan4nqbszv5phxdvaz92zgcim0rw-kwidgetsaddons-5.58.0.drv
  /nix/store/7lh427qc6k15ldjv74hj1nv4ljk620dn-kcodecs-5.58.0.drv
  /nix/store/p9xcgrxck6054c0p89yq88shkjii84bc-kguiaddons-5.58.0.drv
  /nix/store/34s0d4cqxfhbcv0sdbnar745ldqh7rja-kconfigwidgets-5.58.0.drv
  /nix/store/5mfz5fsydzn557yr9xvkdrrkbs6a842s-kitemmodels-5.58.0.drv
  /nix/store/16knw60r4jnbnj0ia5zw02ah04xgz49f-kjobwidgets-5.58.0.drv
  /nix/store/0vrbmb17a1vbdm1sx6v48djfxb7g0r8n-breeze-icons-5.58.0.drv
  /nix/store/w47ihgiifgspa2hly7yr7nrxgs3351v2-kitemviews-5.58.0.drv
  /nix/store/hxj5vqpbwydz0lzaa4fgsilddim77kvm-kiconthemes-5.58.0.drv
  /nix/store/lywfldvrra02wv6mp9xws3gac24xn90h-knotifications-5.58.0.drv
  /nix/store/9wa613w59ia0gdr46726s9l1qw9rwsyg-solid-5.58.0.drv
  /nix/store/aa8p5z73xfxaq00hg94mxryg0l9m05r8-kwallet-5.58.0.drv
  /nix/store/422gf1nqr2w2dxb4fl7bc38g81p2n42d-attica-5.58.0.drv
  /nix/store/x602wpd9nyl8xminy9fd0n7igla1ipf2-sonnet-5.58.0.drv
  /nix/store/rihafn6xpp0j02rcsmbpmsnw8472d7bh-kcompletion-5.58.0.drv
  /nix/store/x8l25sgsk24ba60jyyl95fzrb72ay8yq-ktextwidgets-5.58.0.drv
  /nix/store/zdlq12libn7mg190kyhk7qfw4plcbx2q-kglobalaccel-5.58.0.drv
  /nix/store/v40n7f4c1frhg5q9xl7sdhdawm5chcyf-kxmlgui-5.58.0.drv
  /nix/store/hdax8vlhsh321hyz426mrhinmmv3ncgr-kbookmarks-5.58.0.drv
  /nix/store/pnss0prnn072clxvgyyfp9rvhzy63n0d-kio-5.58.0.drv
  /nix/store/72nnw7dp7hgqxlcffa44hah6mbr9lxgw-kparts-5.58.0.drv
  /nix/store/6jp5q025zaqlhlrgs7q1aia9105bybxz-kactivities-5.58.0.drv
  /nix/store/gdvsc3ywn9d6jhqgwdgdmpp64y7h50vg-qtquickcontrols2-5.12.0.drv
  /nix/store/hw6qbsvwgracda0jg0xzdwhx6wgngd3m-kpackage-5.58.0.drv
  /nix/store/j6qfjhz32bg5xgzg5pfb497xlyiyw1ih-kdeclarative-5.58.0.drv
  /nix/store/l37hj1jvpcylhb6b0wg4h59qx6dzf0sw-kirigami2-5.58.0.drv
  /nix/store/qws5jgyc8g5hy01xim3qwr5rqy8p7xih-kwayland-5.58.0.drv
  /nix/store/gww0y5lmjqdr51nfn4a5b0ibfgq714fv-plasma-framework-5.58.0.drv
  /nix/store/z38ihpr86p9arm5p1s7r64ia8b8crkls-kinit-5.58.0.drv
  /nix/store/7qrjcir80j01adcjhg4w5s40908an062-kded-5.58.0.drv
  /nix/store/a1p62nx1ywaq2n2xb5h2bvyc6psnk4m5-kunitconversion-5.58.0.drv
  /nix/store/mwxchyvicdaby5mwmdvvk2lsg4886vsi-kplotting-5.58.0.drv
  /nix/store/nldchj8bj0chrg0xhr5nwb1ibh96jz3h-kdesignerplugin-5.58.0.drv
  /nix/store/040f04i0iq1xvvzw51xs0cydzz64lh0f-kdelibs4support-5.58.0.drv
  /nix/store/1qjwy8sx37fxqf3fnf18gj8wi4vsy4hi-knotifyconfig-5.58.0.drv
  /nix/store/qd2h099fxk0mqyk3zj10dqvyvhn8wjhj-kpty-5.58.0.drv
  /nix/store/w2wdr67jdcqlrj4izx5vbzz8c1rrnhfv-knewstuff-5.58.0.drv
  /nix/store/inkw3sdwabxfr7hcf2fzhp06wdcczqwl-konsole-19.04.1.drv
building '/nix/store/g22kvv5csgjha330r55dhxqyabfifd1s-qtx11extras-5.12.0.drv' on 'ssh://icvm'...
building '/nix/store/4n6d5l7d4wydmmfl63kyvyycyqbncx69-qtsvg-5.12.0.drv' on 'ssh://icvm'...
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
(repeated forever)

root@xssd:/root# nix-env -f /var/nixpkgs -iA nixUnstable
replacing old 'nix-2.3pre6779_324a5dc9'
installing 'nix-2.3pre6779_324a5dc9'

root@xssd:/root# rehash

root@xssd:/root# nix --version
nix (Nix) 2.3pre6779_324a5dc9

root@xssd:/root# nix-env -f /var/nixpkgs -iA chromium konsole
installing 'chromium-75.0.3770.142'
installing 'konsole-19.04.1'
these derivations will be built:
  /nix/store/g22kvv5csgjha330r55dhxqyabfifd1s-qtx11extras-5.12.0.drv
  /nix/store/4n6d5l7d4wydmmfl63kyvyycyqbncx69-qtsvg-5.12.0.drv
  /nix/store/lqrh0vlwhlryklj69f67iy67sdi9avis-qtdeclarative-5.12.0.drv
  /nix/store/l4nsjzxfpb3akzdk08560dx2qazdy8vd-qtquickcontrols-5.12.0.drv
  /nix/store/9b510cpp2k8rj4xjlzy3r13i2n1b3h8h-qtwayland-5.12.0.drv
  /nix/store/x2q7rpfr0i1iww64bp3r3mlds8mnnprx-hook.drv
  /nix/store/0yl1531z7bzsga74hzimmyvbj7srzf97-kwindowsystem-5.58.0.drv
  /nix/store/kcl9n7hbs0xsf2s554ipwphksad3bi2f-kcoreaddons-5.58.0.drv
  /nix/store/1w4706c2w6mdmz8wi0s8kg341pn948gc-kcrash-5.58.0.drv
  /nix/store/g8sg5ynzaqpkqhwvsvzsdqbk44hv4a6p-karchive-5.58.0.drv
  /nix/store/p0wd2451ywsj496xlhfgz2qjza5lxnq7-ki18n-5.58.0.drv
  /nix/store/854dn5k74zl27hix44mmfchi4713sgz2-kdoctools-5.58.0.drv
  /nix/store/f078jw6vdxyvwvyk5g9ns7pncp8rjlv2-kconfig-5.58.0.drv
  /nix/store/k1rm0ssnybil66qrnfb2jssm5fm74jjw-kdbusaddons-5.58.0.drv
  /nix/store/694gh69q3n45n5wiaqry9dcbs1ckv94y-kservice-5.58.0.drv
  /nix/store/2xjraz3xqi2axhjc067j80jv28hxw98s-kemoticons-5.58.0.drv
  /nix/store/4p7qninjj2w5lbgiisg5wz3n6frpqhxy-kauth-5.58.0.drv
  /nix/store/60w8qan4nqbszv5phxdvaz92zgcim0rw-kwidgetsaddons-5.58.0.drv
  /nix/store/7lh427qc6k15ldjv74hj1nv4ljk620dn-kcodecs-5.58.0.drv
  /nix/store/p9xcgrxck6054c0p89yq88shkjii84bc-kguiaddons-5.58.0.drv
  /nix/store/34s0d4cqxfhbcv0sdbnar745ldqh7rja-kconfigwidgets-5.58.0.drv
  /nix/store/5mfz5fsydzn557yr9xvkdrrkbs6a842s-kitemmodels-5.58.0.drv
  /nix/store/16knw60r4jnbnj0ia5zw02ah04xgz49f-kjobwidgets-5.58.0.drv
  /nix/store/0vrbmb17a1vbdm1sx6v48djfxb7g0r8n-breeze-icons-5.58.0.drv
  /nix/store/w47ihgiifgspa2hly7yr7nrxgs3351v2-kitemviews-5.58.0.drv
  /nix/store/hxj5vqpbwydz0lzaa4fgsilddim77kvm-kiconthemes-5.58.0.drv
  /nix/store/lywfldvrra02wv6mp9xws3gac24xn90h-knotifications-5.58.0.drv
  /nix/store/9wa613w59ia0gdr46726s9l1qw9rwsyg-solid-5.58.0.drv
  /nix/store/aa8p5z73xfxaq00hg94mxryg0l9m05r8-kwallet-5.58.0.drv
  /nix/store/422gf1nqr2w2dxb4fl7bc38g81p2n42d-attica-5.58.0.drv
  /nix/store/x602wpd9nyl8xminy9fd0n7igla1ipf2-sonnet-5.58.0.drv
  /nix/store/rihafn6xpp0j02rcsmbpmsnw8472d7bh-kcompletion-5.58.0.drv
  /nix/store/x8l25sgsk24ba60jyyl95fzrb72ay8yq-ktextwidgets-5.58.0.drv
  /nix/store/zdlq12libn7mg190kyhk7qfw4plcbx2q-kglobalaccel-5.58.0.drv
  /nix/store/v40n7f4c1frhg5q9xl7sdhdawm5chcyf-kxmlgui-5.58.0.drv
  /nix/store/hdax8vlhsh321hyz426mrhinmmv3ncgr-kbookmarks-5.58.0.drv
  /nix/store/pnss0prnn072clxvgyyfp9rvhzy63n0d-kio-5.58.0.drv
  /nix/store/72nnw7dp7hgqxlcffa44hah6mbr9lxgw-kparts-5.58.0.drv
  /nix/store/6jp5q025zaqlhlrgs7q1aia9105bybxz-kactivities-5.58.0.drv
  /nix/store/gdvsc3ywn9d6jhqgwdgdmpp64y7h50vg-qtquickcontrols2-5.12.0.drv
  /nix/store/hw6qbsvwgracda0jg0xzdwhx6wgngd3m-kpackage-5.58.0.drv
  /nix/store/j6qfjhz32bg5xgzg5pfb497xlyiyw1ih-kdeclarative-5.58.0.drv
  /nix/store/l37hj1jvpcylhb6b0wg4h59qx6dzf0sw-kirigami2-5.58.0.drv
  /nix/store/qws5jgyc8g5hy01xim3qwr5rqy8p7xih-kwayland-5.58.0.drv
  /nix/store/gww0y5lmjqdr51nfn4a5b0ibfgq714fv-plasma-framework-5.58.0.drv
  /nix/store/z38ihpr86p9arm5p1s7r64ia8b8crkls-kinit-5.58.0.drv
  /nix/store/7qrjcir80j01adcjhg4w5s40908an062-kded-5.58.0.drv
  /nix/store/a1p62nx1ywaq2n2xb5h2bvyc6psnk4m5-kunitconversion-5.58.0.drv
  /nix/store/mwxchyvicdaby5mwmdvvk2lsg4886vsi-kplotting-5.58.0.drv
  /nix/store/nldchj8bj0chrg0xhr5nwb1ibh96jz3h-kdesignerplugin-5.58.0.drv
  /nix/store/040f04i0iq1xvvzw51xs0cydzz64lh0f-kdelibs4support-5.58.0.drv
  /nix/store/1qjwy8sx37fxqf3fnf18gj8wi4vsy4hi-knotifyconfig-5.58.0.drv
  /nix/store/qd2h099fxk0mqyk3zj10dqvyvhn8wjhj-kpty-5.58.0.drv
  /nix/store/w2wdr67jdcqlrj4izx5vbzz8c1rrnhfv-knewstuff-5.58.0.drv
  /nix/store/inkw3sdwabxfr7hcf2fzhp06wdcczqwl-konsole-19.04.1.drv
building '/nix/store/g22kvv5csgjha330r55dhxqyabfifd1s-qtx11extras-5.12.0.drv' on 'ssh://icvm'...
building '/nix/store/4n6d5l7d4wydmmfl63kyvyycyqbncx69-qtsvg-5.12.0.drv' on 'ssh://icvm'...
copying 1 paths...
copying path '/nix/store/q769ybsa0wqb4kpmncz5bndm4djfb5d9-qtx11extras-5.12.0-dev' from 'ssh://icvm'...
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
file 'nar/2l38315v05wz32virhmdwf3qd556vcrf.nar' does not exist in binary cache 'http://xssd.wg:5000'
(repeated forever)
```

```
root@xssd:/root# ls -l /nix/store/2l38315v05wz32virhmdwf3qd556vcrf*
-rw------- 1 root root 0 2019-07-21 01:25 /nix/store/2l38315v05wz32virhmdwf3qd556vcrf-qtsvg-5.12.0-bin.lock
```

```
# cat /etc/nix/nix.conf
# WARNING: this file is generated from the nix.* options in
# your NixOS configuration, typically
# /etc/nixos/configuration.nix.  Do not edit it!
build-users-group = nixbld
max-jobs = 4
cores = 0
sandbox = true
extra-sandbox-paths =
substituters = http://xssd.wg:5000 http://xhdd.wg:5000 https://cache.nixos.org/
trusted-substituters =
trusted-public-keys = xssd-1:fwMPxKFI/rhC6Zvpks6DD5IWWX96VpKcunN8Vi9TMMA= xhdd-1:QZ/V4lupSRmBX1nO6kzBOE/6Q0y0kL7nplTsH4/Bjdk= cache.nixos.org-1:6NCHdD59X431o0gWypbMrAURkbJ16ZPMQFGspcDShjY=
auto-optimise-store = false
require-sigs = true

trusted-users = root
allowed-users = nix-serve morph root @users

system-features = nixos-test benchmark big-parallel kvm
# Make builders download things, we don't want to push everything
builders-use-substitutes = true

secret-key-files = /var/secrets/nix/secret_key
# Keep source tarballs
keep-outputs = true

# For remote nix builds
trusted-users = at
```

**Comment by ivan** [comment]
The problem seems to have been fixed by `nix-collect-garbage -d`, but maybe this is still a bug?

**Comment by roberth** [comment]
This seems to be a duplicate of #3534, fixed in #4158.
